source/server/listener_manager_impl.cc

```diff
@@ -410,56 +425,126 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
-  if (source_types_array[source_type] != nullptr) {
+  if (source_ports.empty()) {
+    addFilterChainForSourcePorts(source_ips_map[source_ip], 0, filter_chain);
+  } else {
+    for (auto source_port : source_ports) {
+      addFilterChainForSourcePorts(source_ips_map[source_ip], source_port, filter_chain);
+    }
+  }
+}
+
+void ListenerImpl::addFilterChainForSourcePorts(SourcePortsMapSharedPtr& source_ports_map_ptr,
+                                                uint32_t source_port,
+                                                const Network::FilterChainSharedPtr& filter_chain) {
+  if (source_ports_map_ptr == nullptr) {
+    source_ports_map_ptr = std::make_shared<SourcePortsMap>();
+  }
+  auto& source_ports_map = *source_ports_map_ptr;
+
+  if (source_ports_map.find(source_port) != source_ports_map.end()) {
     // If we got here and found already configured branch, then it means that this FilterChainMatch
     // is a duplicate, and that there is some overlap in the repeated fields with already processed
     // FilterChainMatches.
     throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
                                      "overlapping matching rules are defined",
                                      address_->asString()));
   }
-  source_types_array[source_type] = filter_chain;
+  source_ports_map[source_port] = filter_chain;
+}
+
+namespace {
+
+// Template function for creating a CIDR list entry for either source or destination address.
+template <class T>
+std::pair<T, std::vector<Network::Address::CidrRange>> makeCidrListEntry(const std::string& cidr,
+                                                                         const T& data) {
+  std::vector<Network::Address::CidrRange> subnets;
+  if (cidr == EMPTY_STRING) {
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      subnets.push_back(Network::Address::CidrRange::create("0.0.0.0/0"));
+    }
+    if (Network::Address::ipFamilySupported(AF_INET6)) {
+      subnets.push_back(Network::Address::CidrRange::create("::/0"));
+    }
+  } else {
+    subnets.push_back(Network::Address::CidrRange::create(cidr));
+  }
+  return std::make_pair<T, std::vector<Network::Address::CidrRange>>(T(data), std::move(subnets));
 }
 
-void ListenerImpl::convertDestinationIPsMapToTrie() {
+}; // namespace
+
+void ListenerImpl::convertIPsToTries() {
   for (auto& port : destination_ports_map_) {
+    // These variables are used as we build up the destination CIDRs used for the trie.
     auto& destination_ips_pair = port.second;
     auto& destination_ips_map = destination_ips_pair.first;
-    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>> list;
+    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>>
```

Comment:
save a few copies if we add `destination_ips_list.reserve(destination_ips_map.size())`